lib/ProductOpener/Producers.pm

```diff
@@ -0,0 +1,531 @@
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package ProductOpener::Producers;
+
+use utf8;
+use Modern::Perl '2017';
+use Exporter    qw< import >;
+
+use Log::Any qw($log);
+
+
+BEGIN
+{
+	use vars       qw(@ISA @EXPORT @EXPORT_OK %EXPORT_TAGS);
+	@EXPORT = qw();            # symbols to export by default
+	@EXPORT_OK = qw(
+
+		$minion
+
+		&load_csv_or_excel_file
+		&init_columns_fields_match
+		&generate_import_export_columns_groups_for_select2
+		&convert_file
+
+		&import_csv_file_task
+
+					);	# symbols to export on request
+	%EXPORT_TAGS = (all => [@EXPORT_OK]);
+}
+
+use vars @EXPORT_OK ;
+
+use ProductOpener::Config qw/:all/;
+use ProductOpener::Store qw/:all/;
+use ProductOpener::Tags qw/:all/;
+use ProductOpener::Products qw/:all/;
+use ProductOpener::Food qw/:all/;
+use ProductOpener::Ingredients qw/:all/;
+use ProductOpener::Lang qw/:all/;
+use ProductOpener::Display qw/:all/;
+
+use CGI qw/:cgi :form escapeHTML/;
+use URI::Escape::XS;
+use Storable qw/dclone/;
+use Encode;
+use JSON::PP;
+use Time::Local;
+use Data::Dumper;
+use Spreadsheet::CSV();
+use Text::CSV();
+use Minion;
+
+# Minion backend
+
+if (not defined $server_options{minion_backend}) {
+
+	print STDERR "No Minion backend configured in lib/ProductOpener/Config2.pm\n";
+}
+else {
+	print STDERR "Initializing Minion backend configured in lib/ProductOpener/Config2.pm\n";
+	$minion = Minion->new(%{$server_options{minion_backend}});
+}
+
+
+# Load a CSV or Excel file
+
+sub load_csv_or_excel_file($) {
+
+	my $file = shift;	# path and file name
+
+	my $headers_ref;
+	my $rows_ref = [];
+	my $results_ref = { };
+
+	# Spreadsheet::CSV does not like CSV files with a BOM:
+	# Wide character in print at /usr/local/share/perl/5.24.1/Spreadsheet/CSV.pm line 87.
+	#
+
+	# There are many issues with Spreadsheet::CSV handling of CSV files
+	# (depending on whether there is a BOM, encoding, line endings etc.
+	# -> use Spreadsheet::CSV only for Excel files
+	# -> use Text::CSV directly for CSV files
+
+	my $extension = $file;
+	$extension =~ s/^(.*)\.//;
+	$extension = lc($extension);
+
+	if (($extension eq "csv") or ($extension eq "tsv") or ($extension eq "txt")) {
+
+		my $encoding = "UTF-8";
+
+		$log->debug("opening CSV file", { file => $file, extension => $extension }) if $log->is_debug();
+
+		my $csv_options_ref = { binary => 1 , sep_char => "\t" };	# should set binary attribute.
```

Comment:
Do we want to assume that CSV and TSV all use TAB as the separator? As we want to allow producers to upload custom formats, this might vary.

Reply:
In practice most of them will send Excel files. So for CSV we can mandate tabs and Unicode. We want the platform to be as simple as possible.

Reply:
OK